client/send-video.cc

```diff
@@ -118,148 +132,167 @@ int main(int argc, char *argv[]) {
+        pCodecCtxOrig = pFormatCtx->streams[videoStream]->codec;
+        double fps = av_q2d(pFormatCtx->streams[videoStream]->avg_frame_rate);
+        if (fps < 0) {
+            fps = 1.0 / av_q2d(pFormatCtx->streams[videoStream]->codec->time_base);
         }
+        if (verbose) fprintf(stderr, "FPS: %f\n", fps);
 
-        // Free the packet that was allocated by av_read_frame
-        av_free_packet(&packet);
-    }
+        // Find the decoder for the video stream
+        pCodec=avcodec_find_decoder(pCodecCtxOrig->codec_id);
+        if(pCodec==NULL) {
+            fprintf(stderr, "Unsupported codec!\n");
+            return -1; // Codec not found
+        }
+        // Copy context
+        pCodecCtx = avcodec_alloc_context3(pCodec);
+        if(avcodec_copy_context(pCodecCtx, pCodecCtxOrig) != 0) {
+            fprintf(stderr, "Couldn't copy codec context");
+            continue; // Error copying codec context
+        }
 
-    if (interrupt_received) {
-        // Feedback for Ctrl-C, but most importantly, force a newline
-        // at the output, so that commandline-shell editing is not messed up.
-        fprintf(stderr, "Got interrupt. Exiting\n");
-    }
+        // Open codec
+        if(avcodec_open2(pCodecCtx, pCodec, NULL)<0)
+            continue; // Could not open codec
+
+        // Allocate video frame
+        pFrame=av_frame_alloc();
+
+        // Allocate an AVFrame structure
+        pFrameRGB=av_frame_alloc();
+        if(pFrameRGB==NULL)
+            return -1;
+
+        // Determine required buffer size and allocate buffer
+        numBytes=avpicture_get_size(AV_PIX_FMT_RGB24, pCodecCtx->width,
+                                    pCodecCtx->height);
+        buffer=(uint8_t *)av_malloc(numBytes*sizeof(uint8_t));
+
+        // Assign appropriate parts of buffer to image planes in pFrameRGB
+        // Note that pFrameRGB is an AVFrame, but AVFrame is a superset
+        // of AVPicture
+        avpicture_fill((AVPicture *)pFrameRGB, buffer, AV_PIX_FMT_RGB24,
+                       pCodecCtx->width, pCodecCtx->height);
+
+        // initialize SWS context for software scaling
+        sws_ctx = sws_getContext(pCodecCtx->width,
+                                 pCodecCtx->height,
+                                 pCodecCtx->pix_fmt,
+                                 display.width(), display.height(),
+                                 AV_PIX_FMT_RGB24,
+                                 SWS_BILINEAR,
+                                 NULL,
+                                 NULL,
+                                 NULL
+                                 );
+        if (sws_ctx == 0) {
+            fprintf(stderr, "Trouble doing scaling to %dx%d :(\n",
+                    display.width(), display.height());
+            return 1;
+        }
 
-    if (off_z > 0) {
-        display.Clear();
-        display.Send();
-    }
+        signal(SIGTERM, InterruptHandler);
```

Comment:
this is initialized in every loop here. We probably want to pull that out of here.